cpp/src/io/utilities/file_io_utilities.hpp

```diff
@@ -0,0 +1,242 @@
+                                                   rmm::cuda_stream_view stream) = 0;
+
+  /**
+   * @brief Reads into existing device memory.
+   *
+   *  @throws cudf::logic_error on cuFile error
+   *
+   * @param offset Number of bytes from the start
+   * @param size Number of bytes to read
+   * @param dst Address of the existing device memory
+   * @param stream CUDA stream to use, default `rmm::cuda_stream_default`
+   *
+   * @return The number of bytes read
+   */
+  virtual size_t read(size_t offset, size_t size, uint8_t *dst, rmm::cuda_stream_view stream) = 0;
+};
+
+/**
+ * @brief Interface class for cufile output.
+ */
+class cufile_output : public cufile_io_base {
+ public:
+  /**
+   * @brief Writes the data from a device buffer into a file.
+   *
+   *  @throws cudf::logic_error on cuFile error
+   *
+   * @param data Pointer to the buffer to be written into the output file
+   * @param offset Number of bytes from the start
+   * @param size Number of bytes to write
+   */
+  virtual void write(void const *data, size_t offset, size_t size) = 0;
+};
+
+#ifdef CUFILE_FOUND
+
+class cufile_shim;
+
+/**
+ * @brief Class that provides RAII for cuFile file registration.
+ */
+struct cufile_registered_file {
+  void register_handle();
+
+ public:
+  cufile_registered_file(cufile_shim const *shim, std::string const &filepath, int flags)
+    : _file(filepath, flags), shim{shim}
+  {
+    register_handle();
+  }
+
+  cufile_registered_file(cufile_shim const *shim,
+                         std::string const &filepath,
+                         int flags,
+                         mode_t mode)
+    : _file(filepath, flags, mode), shim{shim}
+  {
+    register_handle();
+  }
+
+  auto const &handle() const noexcept { return cf_handle; }
+
+  ~cufile_registered_file();
+
+ private:
+  file_wrapper const _file;
+  CUfileHandle_t cf_handle = nullptr;
+  cufile_shim const *shim  = nullptr;
+};
+
+/**
+ * @brief Adapter for the `cuFileRead` API.
+ *
+ * Exposes APIs to read directly from a file into device memory.
+ */
+class cufile_input_impl final : public cufile_input {
+ public:
+  cufile_input_impl(std::string const &filepath);
+
+  std::unique_ptr<datasource::buffer> read(size_t offset,
+                                           size_t size,
+                                           rmm::cuda_stream_view stream) override;
+
+  size_t read(size_t offset, size_t size, uint8_t *dst, rmm::cuda_stream_view stream) override;
+
+ private:
+  cufile_shim const *shim = nullptr;
+  cufile_registered_file const cf_file;
+};
+
+/**
+ * @brief Adapter for the `cuFileWrite` API.
+ *
+ * Exposes an API to write directly into a file from device memory.
+ */
+class cufile_output_impl final : public cufile_output {
+ public:
+  cufile_output_impl(std::string const &filepath);
+
+  void write(void const *data, size_t offset, size_t size) override;
+
+ private:
+  cufile_shim const *shim = nullptr;
```

Comment:
I'd rather not reuse the shim pointer here. Doesn't feel right from the OOP/testing perspective; both _impl classes and `cufile_registered_file` use the shim (differently). Let me know if you feel otherwise.